sources/hba_reader.c

```diff
@@ -0,0 +1,397 @@
+
+/*
+ * Odyssey.
+ *
+ * Scalable PostgreSQL connection pooler.
+ */
+
+#include <arpa/inet.h>
+#include <kiwi.h>
+#include <machinarium.h>
+#include <odyssey.h>
+
+enum { OD_LLOCAL,
+       OD_LHOST,
+       OD_LHOSTSSL,
+       OD_LHOSTNOSSL,
+       OD_LALL,
+       OD_LSAMEUSER,
+       OD_LALLOW,
+       OD_LDENY,
+};
+
+static od_keyword_t od_hba_keywords[] = {
+	/* connection types */
+	od_keyword("local", OD_LLOCAL),
+	od_keyword("host", OD_LHOST),
+	od_keyword("hostssl", OD_LHOSTSSL),
+	od_keyword("hostnossl", OD_LHOSTNOSSL),
+	/* db/user */
+	od_keyword("all", OD_LALL),
+	od_keyword("sameuser", OD_LSAMEUSER),
+	/* auth type */
+	od_keyword("allow", OD_LALLOW),
+	od_keyword("trust", OD_LALLOW),
+	od_keyword("deny", OD_LDENY),
+	od_keyword("reject", OD_LDENY),
+};
+
+static void od_hba_reader_error(od_config_reader_t *reader, char *msg)
+{
+	od_errorf(reader->error, "%s:%d %s", reader->config_file,
+		  reader->parser.line, msg);
+}
+
+static int od_hba_parser_next(od_parser_t *parser, od_token_t *token)
+{
+	/* try to use backlog */
+	if (parser->backlog_count > 0) {
+		*token = parser->backlog[parser->backlog_count - 1];
+		parser->backlog_count--;
+		return token->type;
+	}
+	/* skip white spaces and comments */
+	for (;;) {
+		while (parser->pos < parser->end && isspace(*parser->pos)) {
+			if (*parser->pos == '\n')
+				parser->line++;
+			parser->pos++;
+		}
+		if (od_unlikely(parser->pos == parser->end)) {
+			token->type = OD_PARSER_EOF;
+			return token->type;
+		}
+		if (*parser->pos != '#')
+			break;
+		while (parser->pos < parser->end && *parser->pos != '\n')
+			parser->pos++;
+		if (parser->pos == parser->end) {
+			token->type = OD_PARSER_EOF;
+			return token->type;
+		}
+		parser->line++;
+	}
+
+	/* symbols */
+	if (*parser->pos != '\"' && ispunct(*parser->pos)) {
+		token->type = OD_PARSER_SYMBOL;
+		token->line = parser->line;
+		token->value.num = *parser->pos;
+		parser->pos++;
+		return token->type;
+	}
+
+	if (isalnum(*parser->pos)) {
+		token->type = OD_PARSER_KEYWORD;
+		token->line = parser->line;
+		token->value.string.pointer = parser->pos;
+		while (parser->pos < parser->end && *parser->pos != ',' &&
+		       (isalnum(*parser->pos) || ispunct(*parser->pos)))
+			parser->pos++;
+		token->value.string.size =
+			parser->pos - token->value.string.pointer;
+		return token->type;
+	}
```

Comment:
so this is keyword parsing, why is it differs from od_parser_next keyword parsing?

```
        /* keyword */
        if (isalpha(*parser->pos)) {
                token->type = OD_PARSER_KEYWORD;
                token->line = parser->line;
                token->value.string.pointer = parser->pos;
                while (parser->pos < parser->end &&
                       (*parser->pos == '_' || isalpha(*parser->pos) ||
                        isdigit(*parser->pos)))
                        parser->pos++;
                token->value.string.size =
                        parser->pos - token->value.string.pointer;
                return token->type;
        }
```